api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java

```diff
@@ -219,21 +220,17 @@ public Patient getPatient(Integer patientId) throws APIException {
 	@Override
 	@Transactional(readOnly = true)
 	public Patient getPatientOrPromotePerson(Integer patientOrPersonId) {
```

Comment:
@dkayiwa I was able to reproduce the issue and see the same errors that you reporting with my previous fix.  Here is my alternative solution, which fixes the issue for me.  Rather than try to load the patient, and then fall back to loading the person, this first loads the person, and checks to see if that person represents a valid patient or not.  This seems like the more correct behavior anyway, given our inheritance model.

To get this working in the refapp, I also needed to apply a related fix to idgen, which I raised a separate PR for, though you could also just disable the idgen module to test this standalone.

Thoughts on this approach?  This is a PR against the 2.3.x branch, which if accepted, we can up-port as needed.

Reply:
This generally seems correct, but I'd be leery of backporting it if it's just to fix the bug reported in TRUNK-5977... does this manifest itself it somewhere, or is there a reason this bug is more insidious than on might believe. It seems like a bit of an edge case (registering an existing provider or other relationship person as a patient)?

Another option would be to just change this to catch the other exception thrown in 2.3.x, and then make the larger change in 2.4.x?

Reply:
@mogoodrich - as previous PR comments (now marked as resolved) with @dkayiwa discovered, simply catching the other exception wasn't enough.  There were other exceptions that occurred.

The entire reason we are working on this and coding it against 2.3.x is because this is a bug that is manifesting in our soon-to-go-to-production distribution in Rwanda that is running 2.3.2, and we need the fix to go into 2.3.3.

Reply:
I think this is a more common use case than the ticket might have you believe - as in the Rwanda distribution it is common to create relationships and providers, including CHWs, on the patient dashboard, and then later if/when they become patients to promote these persons to patients via the process described here.

Reply:
Okay, fair enough... once we merge this we may actually want to upgrade the PIH EMR to 2.3.x to test...

Reply:
OK, will wait for approval or change requests until Monday and then look to commit.

Reply:
@mseaton your solution is much better than the original code! 👍